pkg/netlink/netlink_watcher.go

```diff
@@ -0,0 +1,2088 @@
+		return true
+	}
+
+	return false
+}
+
+// parseNeigh parses the neigh
+func parseNeigh(nm []neighIPStruct, v string) neighList {
+	var NL neighList
+	for _, ND := range nm {
+		var ns neighStruct
+		ns.Neigh0.Type = OTHER
+		ns.VrfName = v
+		if ND.Dev != "" {
+			vrf, _ := vn.LinkByName(ND.Dev)
+			ns.Neigh0.LinkIndex = vrf.Attrs().Index
+		}
+		if ND.Dst != "" {
+			ipnet := &net.IPNet{
+				IP: net.ParseIP(ND.Dst),
+			}
+			ns.Neigh0.IP = ipnet.IP
+		}
+		if len(ND.State) != 0 {
+			ns.Neigh0.State = getState(ND.State[0])
+		}
+		if ND.Lladdr != "" {
+			ns.Neigh0.HardwareAddr, _ = net.ParseMAC(ND.Lladdr)
+		}
+		if ND.Protocol != "" {
+			ns.Protocol = ND.Protocol
+		}
+		//	ns  =  neighborAnnotate(ns)   /* Need InfraDB to finish for fetching LB/BP information */
+		ns.Key = neighKey{VrfName: v, Dst: ns.Neigh0.IP.String(), Dev: ns.Neigh0.LinkIndex}
+		if preFilterNeighbor(ns) {
+			NL.NS = append(NL.NS, ns)
+		}
+	}
+	return NL
+}
+
+// getNeighborRoutes gets the nighbor routes
+func getNeighborRoutes() []routeCmdInfo { // []map[string]string{
+	// Return a list of /32 or /128 routes & Nexthops to be inserted into
+	// the routing tables for Resolved neighbors on connected subnets
+	// on physical and SVI interfaces.
+	var neighborRoutes []routeCmdInfo // []map[string]string
+	for _, N := range latestNeighbors {
+		if N.Type == PHY || N.Type == SVI || N.Type == VXLAN {
+			vrf, _ := infradb.GetVrf(N.VrfName)
+			table := int(*vrf.Metadata.RoutingTable[0])
+
+			//# Create a special route with dst == gateway to resolve
+			//# the nexthop to the existing neighbor
+			R0 := routeCmdInfo{Type: routeTypeNeighbor, Dst: N.Neigh0.IP.String(), Protocol: "ipu_infra_mgr", Scope: "global", Gateway: N.Neigh0.IP.String(), Dev: NameIndex[N.Neigh0.LinkIndex], VRF: vrf, Table: table}
+			neighborRoutes = append(neighborRoutes, R0)
+		}
+	}
+	return neighborRoutes
+}
+
+// readNeighbors reads the nighbors
+func readNeighbors(v *infradb.Vrf) {
+	var N neighList
+	var err error
+	var Nb string
+	if v.Spec.Vni == nil {
+		/* No support for "ip neighbor show" command in netlink library Raised ticket https://github.com/vishvananda/netlink/issues/913 ,
+		   so using ip command as WA */
+		Nb, err = nlink.ReadNeigh(ctx, "")
+	} else {
+		Nb, err = nlink.ReadNeigh(ctx, path.Base(v.Name))
+	}
+	if len(Nb) != 3 && err == nil {
+		N = cmdProcessNb(Nb, v.Name)
+	}
+	addNeigh(N)
+}
+
+// preFilterMac filter the mac
+func preFilterMac(f *FdbEntryStruct) bool {
+	// TODO m.nexthop.dst
+	if f.VlanID != 0 || (f.Nexthop.Dst != nil && !f.Nexthop.Dst.IsUnspecified()) {
+		log.Printf("netlink: %d vlan \n", len(f.Nexthop.Dst.String()))
+		return true
+	}
+	return false
+}
+
+// cmdProcessRt process the route command
+func cmdProcessRt(v *infradb.Vrf, r string, t int) routeList {
+	var RouteData []routeCmdInfo
+	if len(r) <= 3 {
+		log.Println("netlink: Error in the cmd:", r)
+		var route routeList
+		return route
+	}
+	CPs := strings.Split(r[2:len(r)-3], "},{")
```

Comment:
is `r` json? Unmarshall?

Reply:
r is the raw output of the linux command, in the format of [{},....,], which is later unmarshalled

Reply:
I don't like these string manipulations. It is too easy to get an error (for example, what if a whitespace in some versions is added in between of brace and comma?)
You can try to unmarshal the result into an array and see how many elements you have there and subsequently use maps in that array to work further (not passing strings, they are unreliable).
You also repeat this procedure in different places. Search by "},{" shows 4 places

Reply:
Done